components/omnibox/browser/commander_provider.cc

```diff
@@ -0,0 +1,101 @@
+#include "brave/components/commander/common/commander_url.h"
+#include "brave/components/commander/common/constants.h"
+#include "components/omnibox/browser/autocomplete_match.h"
+#include "components/omnibox/browser/autocomplete_match_type.h"
+#include "components/omnibox/browser/autocomplete_provider_client.h"
+#include "components/omnibox/browser/autocomplete_provider_listener.h"
+
+namespace commander {
+CommanderProvider::CommanderProvider(AutocompleteProviderClient* client,
+                                     AutocompleteProviderListener* listener)
+    : AutocompleteProvider(AutocompleteProvider::TYPE_BRAVE_COMMANDER) {
+  if (listener) {
+    AddListener(listener);
+  }
+
+  if (auto* delegate = commander::CommanderFrontendDelegate::Get()) {
+    delegate->AddObserver(this);
+  }
+}
+
+CommanderProvider::~CommanderProvider() {
+  if (auto* delegate = commander::CommanderFrontendDelegate::Get()) {
+    delegate->RemoveObserver(this);
+  }
+}
+
+void CommanderProvider::Start(const AutocompleteInput& input,
+                              bool minimal_changes) {
+  last_input_ = input.text();
+
+  // The Omnibox is setup before the commnander when the browser launches, so
+  // it's possible we haven't been added as an observer yet.
+  if (auto* delegate = commander::CommanderFrontendDelegate::Get()) {
+    if (!IsInObserverList()) {
+      delegate->AddObserver(this);
+    }
+  }
+  OnCommanderUpdated();
+}
+
+void CommanderProvider::OnCommanderUpdated() {
+  auto* delegate = commander::CommanderFrontendDelegate::Get();
+  if (!delegate) {
+    return;
+  }
+
+  matches_.clear();
+
+  if (!last_input_.starts_with(commander::kCommandPrefix)) {
+    return;
+  }
+
+  int rank = 10000;
+  const auto& items = delegate->GetItems();
+  for (uint32_t i = 0; i < items.size(); ++i) {
+    const auto& option = items[i];
+    AutocompleteMatch match(this, rank--, false,
+                            AutocompleteMatchType::BOOKMARK_TITLE);
+    // This is neat but it would be nice if we could always show it instead of
+    // only when we have a result selected.
+    match.contents = option.annotation;
+    match.additional_text = delegate->GetPrompt();
+    if (!option.annotation.empty()) {
+      match.contents_class = {
+          ACMatchClassification(0, ACMatchClassification::DIM)};
+    }
+    match.description =
+        commander::kCommandPrefix + std::u16string(u" ") + option.title;
+    match.allowed_to_be_default_match = true;
+    match.swap_contents_and_description = true;
+    // We don't want to change the prompt at all while the user is going through
+    // their options.
+    match.fill_into_edit = last_input_;
+    match.destination_url =
+        commander::GetCommandURL(i, delegate->GetResultSetId());
+    match.description_class = {
+        ACMatchClassification(0, ACMatchClassification::DIM),
+        ACMatchClassification(2, ACMatchClassification::MATCH)};
+    matches_.push_back(match);
+  }
+
+  NotifyListeners(true);
```

Comment:
nit:

```suggestion
  NotifyListeners(/*updated_matches=*/ true);
```

Reply:
Hahaha I always forget because VSCode has inline hints:
![image](https://user-images.githubusercontent.com/7678024/217113569-98d55e08-add0-4ca0-9fd7-fde63f10aa16.png)

Reply:
Wow, that's slick! What extension is doing that?